### DISCUSSION TITLE
There are too much ways to define ORM mapping object and methods for the same purpose.

### DISCUSSION THREAD

**Discussion by wangxin688** [sqlalchemy/sqlalchemy#9320] [Usage Questions] [answered]
First, I must say sqlalchemy is a very powerful and flexible ORM tools than I ever seen and used. Appreciate for your hard work for a long time development and support.

The most challenging for me and our team is that sqla is too much power and flexible. There are so many ways to define object and CRUD. 

And due to the history reason , a lots of feature need compatibility with previous versions. Since sqlalchemy is upgrading to 2.0 which bring a lot new awesome features. especially for type-hint.  I think it's a milestone,right? So, is there any plan to unify the API style to reduce diversity?  

Altough if we start with different perspectives and different people's views, each man has his own taste. But it would be better you can provider some best practices from developer's side,  the will be more helpful for new bird or some one is refreshing his project. 

I hope this won't bother you.

Thanks a lot.

**Comment by CaselIT** [comment]
Hi,

I'm not sure what you are asking here. Have you taken a look at the current v2 docs? There is a single ORM mapping style used there, unless you go look in the "legacy" style document, that indeed has the old styles.

I'm not sure how it could be improved, unless you mean that v2 should have removed all previous supported versions.

Regarding the possibility of removing currently supported style, I think that the 1.4 dataclass style makes sense to deprecate and eventually remove in soonish (say 2.2). For the other style I don't think it's feasible to do without causing a lot of pain to the users.

**Comment by zzzeek** [comment, marked as answer]
hi -

I get this kind of critique a lot and I can only assume it comes from people who have been with the project a long time and have seen the changes in styles that are offered.   Because, if you actually make a run through the documentation, you would find that there is very **little** diversity in how to make a mapping in 99.9% of all documentation examples; it uses the latest [type-mapped declarative style](https://docs.sqlalchemy.org/en/20/orm/quickstart.html) that you see in the quickstart and in the [unified tutorial](https://docs.sqlalchemy.org/en/20/tutorial/metadata.html#using-orm-declarative-forms-to-define-table-metadata), which is 1. DeclarativeBase 2. mapped_column()  3. Mapped[] annotations (eventually, most examples should use annotations).      That's the style, that's what we want you to do, and it should be clear.   It is also very likely that we will further push this style to standardize on 4. MappedAsDataclass as part of the base mapping, once this feature is more mature.


so if you look at the docs as though you were just coming to SQLAlchemy, you would not see much "diversity", there's basically a single style noted everywhere, with the exception of the [mapping styles](https://docs.sqlalchemy.org/en/20/orm/mapping_styles.html#orm-mapping-styles) reference documentation.   In this section, there are reference examples on a few variations of style that our userbase has made it abundantly clear they want; these variations are:   1. DeclarativeBase vs registry.mapped()   2.   `__tablename__` vs. `__table__` and 3. declarative vs. "imperative".    

Option 1. is cosmetic, decorator vs. base class.   The base class approach is better because the class can be regulated by a metaclass, and we dont discuss the decorator approach much.       

Option 2, `__table__` is an important API feature since you can map classes to joins and select() statements, it makes sense that you can also map to an explicit `Table` object, and people do use that approach.    So `__tablename__` vs. `__table__` makes sense as first class API

Option 3:  I wanted to remove "imperative" mappings entirely however there was a very clear response from the community that a lot of people greatly prefer imperative mappings, so instead I have unified the internal mechanics for declarative and imperative mappings, such that you get the same class behaviors from both styles.   Imperative mappings are only discussed in this one reference document, by now should be removed from 100% of examples elsewhere throughout the documentation, it is not our recommended style, lots of people swear by it, it does not get in anyones way being there.

Now when you say "diversity", you might be thinking of the declarative style where one uses `Column()` objects and not `mapped_column()`.  Again, I would point to the documentation.  You should not find **any** examples of that mapping style outside of "migration from 1.4" documentation;  there are likely examples of it still in the examples/ folder and on the wiki, but those examples should all be changed; PRs welcome for examples/ that are still on the old style.   if you are confused that you can make declarative mappings with either `mapped_column()` or `Column()`, refer to the 2.0 documentation which should make it very clear that the `Column` style is **legacy**.

As far as legacy features still being around, SQLAlchemy's declarative style with Column goes back to 2009 or so and there are millions of applications using it, as there are millions of applications using ``session.query()``.  There is no need or advantage to actually removing the compatibility of these APIs from the library; they are not in contradiction with the current systems.  Query basically lives inside of `query.py` and otherwise has virtually no footprint in the rest of the source code; `Column` in declarative shares most of the same logic with that of `mapped_column()`.     Legacy APIs and behaviors should not be seen as "diversity", that's not what those are.

**Reply by wangxin688** [reply]
Thanks for you guys reply. I had upgrade our project's ORM version to the newest version.As you said, dataclass style will be the recommended way. I will read the docs for more details and change the style it.

**Reply by zzzeek** [reply]
update: I have located docs that still have legacy styles, like hybrid.py.    I am updating hybrid now.  there's probably some more still in there.  the main docs though should be all mapped_column() by now.
